src/main/scala/rise/core/types/Kinds.scala

```diff
@@ -1,39 +1,77 @@
+sealed trait Kind[+T, +I, KI <: Kind.Identifier] {
   def name :String
 }
 
+
 object Kind {
-  trait Identifier {
-    def name: String
+  trait Identifier { def name: String }
+  case class IType(id : TypeIdentifier) extends Identifier { def name : String = id.name }
+  case class IDataType(id : DataTypeIdentifier) extends Identifier { def name : String = id.name }
+  case class INat(id : NatIdentifier) extends Identifier { def name : String = id.name }
+  case class IAddressSpace(id : AddressSpaceIdentifier) extends Identifier { def name : String = id.name }
+  case class INatToNat(id : NatToNatIdentifier) extends Identifier { def name : String = id.name }
+  case class INatToData(id : NatToDataIdentifier) extends Identifier { def name : String = id.name }
+  case class IMatrixLayout(id : MatrixLayoutIdentifier) extends Identifier { def name : String = id.name }
+  case class IFragmentKind(id : FragmentKindIdentifier) extends Identifier { def name : String = id.name }
+  case class INatCollection(id : NatCollectionIdentifier) extends Identifier { def name : String = id.name }
+
+  def idName[T, I, KI <: Kind.Identifier](kind : Kind[T, I, KI], i : I) : String = kind match {
+    case TypeKind => i.asInstanceOf[TypeIdentifier].name
+    case DataKind => i.asInstanceOf[DataTypeIdentifier].name
+    case NatKind => i.asInstanceOf[NatIdentifier].name
+    case AddressSpaceKind => i.asInstanceOf[AddressSpaceIdentifier].name
+    case NatToNatKind => i.asInstanceOf[NatToNatIdentifier].name
+    case NatToDataKind => i.asInstanceOf[NatToDataIdentifier].name
+    case NatCollectionKind => i.asInstanceOf[NatCollectionIdentifier].name
+  }
+
+  def toIdentifier[T, I, KI <: Kind.Identifier](kind : Kind[T, I, KI], i : I) : KI = kind match {
+    case TypeKind => IType(i.asInstanceOf[TypeIdentifier]).asInstanceOf[KI]
+    case DataKind => IDataType(i.asInstanceOf[DataTypeIdentifier]).asInstanceOf[KI]
+    case NatKind => INat(i.asInstanceOf[NatIdentifier]).asInstanceOf[KI]
+    case AddressSpaceKind => IAddressSpace(i.asInstanceOf[AddressSpaceIdentifier]).asInstanceOf[KI]
+    case NatToNatKind => INatToNat(i.asInstanceOf[NatToNatIdentifier]).asInstanceOf[KI]
+    case NatToDataKind => INatToData(i.asInstanceOf[NatToDataIdentifier]).asInstanceOf[KI]
+    case NatCollectionKind => INatCollection(i.asInstanceOf[NatCollectionIdentifier]).asInstanceOf[KI]
+  }
+
+  def fromIdentifier[T, I, KI <: Kind.Identifier](kind : Kind[T, I, KI], i : KI) : I = kind match {
+    case TypeKind => i.asInstanceOf[IType].id.asInstanceOf[I]
+    case DataKind => i.asInstanceOf[IDataType].id.asInstanceOf[I]
+    case NatKind => i.asInstanceOf[INat].id.asInstanceOf[I]
+    case AddressSpaceKind => i.asInstanceOf[IAddressSpace].id.asInstanceOf[I]
+    case NatToNatKind => i.asInstanceOf[INatToNat].id.asInstanceOf[I]
+    case NatToDataKind => i.asInstanceOf[INatToData].id.asInstanceOf[I]
+    case NatCollectionKind => i.asInstanceOf[INatCollection].id.asInstanceOf[I]
   }
 }
```

Comment:
Would it not save on some casting to match on the identifier instead of the kind object?

Reply:
We could do that, but then the match wouldn't be closed (since `I` is not a closed type). If Scala's type system was a little bit more powerful, it could infer that the match on `kind` uniquely determines `I`, and thus the cast wouldn't be necessary. This might be a thing in Scala 3?